eth/filters/bor_filter.go

```diff
@@ -0,0 +1,154 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package filters
+
+import (
+	"context"
+	"errors"
+
+	"github.com/maticnetwork/bor/common"
+	"github.com/maticnetwork/bor/core/types"
+	"github.com/maticnetwork/bor/ethdb"
+	"github.com/maticnetwork/bor/rpc"
+)
+
+// BorBlockLogsFilter can be used to retrieve and filter logs.
+type BorBlockLogsFilter struct {
+	backend Backend
+
+	db        ethdb.Database
+	addresses []common.Address
+	topics    [][]common.Hash
+
+	block      common.Hash // Block hash if filtering a single block
+	begin, end int64       // Range interval if filtering multiple blocks
+}
+
+// NewBorBlockLogsRangeFilter creates a new filter which uses a bloom filter on blocks to
+// figure out whether a particular block is interesting or not.
+func NewBorBlockLogsRangeFilter(backend Backend, begin, end int64, addresses []common.Address, topics [][]common.Hash) *BorBlockLogsFilter {
+	// Create a generic filter and convert it into a range filter
+	filter := newBorBlockLogsFilter(backend, addresses, topics)
+	filter.begin = begin
+	filter.end = end
+
+	return filter
+}
+
+// NewBorBlockLogsFilter creates a new filter which directly inspects the contents of
+// a block to figure out whether it is interesting or not.
+func NewBorBlockLogsFilter(backend Backend, block common.Hash, addresses []common.Address, topics [][]common.Hash) *BorBlockLogsFilter {
+	// Create a generic filter and convert it into a block filter
+	filter := newBorBlockLogsFilter(backend, addresses, topics)
+	filter.block = block
+	return filter
+}
+
+// newBorBlockLogsFilter creates a generic filter that can either filter based on a block hash,
+// or based on range queries. The search criteria needs to be explicitly set.
+func newBorBlockLogsFilter(backend Backend, addresses []common.Address, topics [][]common.Hash) *BorBlockLogsFilter {
+	return &BorBlockLogsFilter{
+		backend:   backend,
+		addresses: addresses,
+		topics:    topics,
+		db:        backend.ChainDb(),
+	}
+}
+
+// Logs searches the blockchain for matching log entries, returning all from the
+// first block that contains matches, updating the start of the filter accordingly.
+func (f *BorBlockLogsFilter) Logs(ctx context.Context) ([]*types.Log, error) {
+	// If we're doing singleton block filtering, execute and return
+	if f.block != (common.Hash{}) {
+		receipt, err := f.backend.GetBorBlockReceipt(ctx, f.block)
+		if err != nil {
+			return nil, err
+		}
+
+		if receipt == nil {
+			return nil, errors.New("unknown block")
+		}
+		return f.borBlockLogs(ctx, receipt)
+	}
+
+	// Figure out the limits of the filter range
+	header, _ := f.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
+	if header == nil {
+		return nil, nil
+	}
+	head := header.Number.Uint64()
+
+	if f.begin == -1 {
+		f.begin = int64(head)
+	}
+
+	// adjust begin for sprint
+	f.begin = nextSprintEnd(f.begin)
```

Comment:
It shouldn't be the `previousSprintEnd`?

Reply:
This function name is confusing. It sounds like it gets the sprint end of the range, and nothing to do with previous or next sprint is that. If it gets 100 as a number, it will return the closest end of 128.

Reply:
@nachomazzara @imyourm8 Sorry for the confusion. It's `currentSprintEnd` actually. For `number = 128`, it will return  128, but for 100 it will return 128 as @imyourm8 mentioned. I will make the change.

Reply:
It means that what happened or is expected on block 100 in Ethereum will be covered by a bor receipt at the beginning of 128 in Matic, right? 

I was wondering if it will be a common use case to set the begin block to get the current sprint start instead of the end (with 100, get 64 instead of 128)

Reply:
Fixed.